File: typescript/projects/Rylogic.TextAligner/src/align/align_group.ts
import {AlignPattern} from './align_pattern';

/** Represents a set of patterns that all align together */
export class AlignGroup
{
	// WARNING: this type is loaded from JSON which doesn't call the constructor.
	// That means functions and properties will be 'undefined'
	constructor(name: string = "", leading_space: number = 0, ...patterns:AlignPattern[])
	{
		this.name = name;
		this.leading_space = leading_space;
		this.patterns = patterns || [];
	}

	/** The name of the group */
	public readonly name :string;

	/** The number leading whitespaces the group should have */
	public readonly leading_space: number;

	/** Patterns belonging to the align group */
	public readonly patterns: AlignPattern[];
}

File: projects/Rylogic.TextAligner/src/Align/AlignGroup.cs
using System.Collections.ObjectModel;
using System.Linq;
using System.Xml.Linq;
using Rylogic.Extn;

namespace Rylogic.TextAligner
{
	/// <summary>Represents a set of patterns that all align together</summary>
	public class AlignGroup
	{
		public AlignGroup()
		{
			Name = string.Empty;
			LeadingSpace = 1;
			Patterns = new ObservableCollection<AlignPattern>();
		}
		public AlignGroup(string name, int leading_space, params AlignPattern[] patterns)
			:this()
		{
			Name = name;
			LeadingSpace = leading_space;
			Patterns.AddRange(patterns);
		}
		public AlignGroup(XElement node)
			:this()
		{
			Name = node.Element(nameof(Name)).As<string>();
			LeadingSpace = node.Element(nameof(LeadingSpace)).As<int>();
			Patterns.AddRange(node.Elements(nameof(Patterns), nameof(AlignPattern)).Select(x => x.As<AlignPattern>()));
		}
		public XElement ToXml(XElement node)
		{
			node.Add2(nameof(Name), Name, false);
			node.Add2(nameof(LeadingSpace), LeadingSpace, false);
			node.Add2(nameof(Patterns), nameof(AlignPattern), Patterns, false);
			return node;
		}

		/// <summary>The name of the group</summary>
		public string Name { get; set; }

		/// <summary>The number leading white spaces the group should have</summary>
		public int LeadingSpace { get; set; }

		/// <summary>The patterns belonging to the group</summary>
		public ObservableCollection<AlignPattern> Patterns { get; set; }

		/// <summary></summary>
		public override string ToString() => Name;
	}
}
